[AIRFLOW-3924] Fix try number for alert emails

TheZepto opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3924

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Fix for alert emails sent via email_alert() to have the correct try number
in the body of the text.

Alert emails are being sent after the state of the Task Instance
is changed from State.RUNNING. This causes self.try_number to be
incremented for the alert email and the first email being sent with
`Try 2 of 1`.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Add a test to ensure the first email sent says `Try 1`.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

ashb commented on airflow/models/__init__.py:
```diff
@@ -1730,6 +1730,8 @@ def email_alert(self, exception):
         # not 0-indexed lists (i.e. Try 1 instead of
         # Try 0 for the first attempt).
+        # This function is called after changing the state
+        # from State.RUNNING so need to subtract 1 from self.try_number.
```
Could you move this comment to line 1720 instead of here?

TheZepto replied:
No worries. I've moved it.

feng-tao replied:
@TheZepto I am not very familiar with this part of the code. But based on this line of comment(https://github.com/apache/airflow/blob/master/airflow/models/__init__.py#L1523), it seems the try_number hasn't increased yet. Do you happen to know which line increases the try number causing the issue?

TheZepto replied:
My understanding is that this comment "Since this function is called only when the TI state is running, try_number contains the current try_number (not the next)." is saying that with a state of running try_number returns the current try. However, changing the state causes try_number to return the next try_number (https://github.com/apache/airflow/blob/master/airflow/models/__init__.py#L706). The state is changed to up_for_retry or failed prior to calling email_alert.

I'm not sure why the behaviour of try_number is to be incremented whenever the state is not equal to running.

feng-tao replied:
thanks.

TheZepto committed:
AIRFLOW-3924 Fix try number in alert emails

Alert emails sent via email_alert() have the correct try number
in the body of the text.
Add a test to ensure the first email sent says `Try 1`.
